### PR TITLE
add __delattr__ to DictConfig

### LINKED ISSUE #682: support __delattr__
```python
from omegaconf import OmegaConf

a = OmegaConf.create()
a.b = {}
a.b["c"] = 1
del a.b["c"]   # works

a.b.c = 1
del a.b.c  # fails
```

**Describe the solution you'd like**
It seems counter-intuitive that this type of deletion is not supported.

### PR DESCRIPTION
**Pull request by jieru-hu**
Closes #682

### CODE REVIEW

**Review by Jasha10 — APPROVED** [review]
Looks good, nice work!

File: omegaconf/dictconfig.py [hunk 1]
```diff
@@ -372,8 +372,21 @@ def __getitem__(self, key: DictKeyType) -> Any:
...
 
+    def __delattr__(self, key: str) -> None:
+        """
+        Allow deleting dictionary values as attributes
+        :param key:
+        :return:
+        """
+        if key == "__name__":
+            raise AttributeError()
+        self._delkey(key)
```
Comment (omry) [inline omegaconf/dictconfig.py:383]: Deleting from an object raises an AttributeError.
When deleting an attribute from a DictConfig backed by a Structured Config I think we should raise ConfigAttributeError.

**Comment by jieru-hu** [comment]
>what do you need special treatment for "__name__" ?


Yea, I'm not super sure about this one - I added this case to be  consistent with the `__getattr__` [behavior](https://github.com/omry/omegaconf/blob/master/omegaconf/dictconfig.py#L341-L348) which seems to make sense. I can  remove it if this is not needed.

**Comment by omry** [comment]
> > what do you need special treatment for "**name**" ?
> 
> Yea, I'm not super sure about this one - I added this case to be consistent with the `__getattr__` [behavior](https://github.com/omry/omegaconf/blob/master/omegaconf/dictconfig.py#L341-L348) which seems to make sense. I can remove it if this is not needed.

1. Yes, remove it. I don't want to add logic without a very clear reason.
2. Can you respond on the actual comment so it's easier to respond in context?

**Review by omry — CHANGES_REQUESTED** [review]
See previous comments.

File: omegaconf/dictconfig.py (hunk 1 above)
Comment (omry) [reply omegaconf/dictconfig.py:381]: Did I miss your response to this?
I am not seeing a reply, and I am not seeing this addressed either.

Comment (Jasha10) [reply omegaconf/dictconfig.py]: FYI @jieru-hu I think that in this case the right way to check if a DictConfig is backed by a Structured Config would be to evaluate `is_structured_config(self._metadata.object_type)`.

Comment (jieru-hu) [reply omegaconf/dictconfig.py]: Sorry I missed this one! I've updated the PR to address this comment.

Comment (Jasha10) [reply omegaconf/dictconfig.py]: Oh, please disregard my previous comment, what you've got now looks good (raising ConfigAttributeError for both structured and non-structured)

**Review by Jasha10 — APPROVED** [review]

File: tests/test_readonly.py [hunk 2]
```diff
@@ -51,12 +51,6 @@
             raises(ReadonlyConfigError, match="a"),
             id="dict_delitem",
         ),
-        param(
-            {"a": 10},
-            lambda c: c.__delattr__("a"),
-            raises(ReadonlyConfigError, match="a"),
-            id="dict_delattr",
-        ),
```
Comment (omry) [inline tests/test_readonly.py]: This one actually makes sense, why did you remove it?

**Review by omry — COMMENTED** [review]
Since Python in general allow deleting of fields from objects with delattr, I think we should allow it for DictConfig in struct mode and for DictConfig backed by a Structured Config. (if they are not readonly).

File: tests/test_errors.py [hunk 3]
```diff
@@ -199,28 +199,6 @@ def finalize(self, cfg: Any) -> None:
...
-        Expected(
-            create=lambda: OmegaConf.create({"foo": "bar"}),
-            op=lambda cfg: cfg.__delattr__("not_found"),
-            key="not_found",
-            exception_type=ConfigKeyError,
-            msg="Key not found: 'not_found'",
-        ),
-        id="dict:delattr_invalid",
-    ),
-    param(
```
Comment (Jasha10) [inline tests/test_errors.py]: Since `__delattr__` is no longer just an alias for `__delitem__`, does it make sense to put some of these tests back in?

Comment (jieru-hu) [reply tests/test_errors.py]: Yes - I think I've added back the tests that still apply. Now we allow deleting attributes from struct and structured,  some of the old tests (in the first commit) in `test_errors` does not apply to `__delattr__` any more since they are expecting the call to fail.

Comment (jieru-hu) [reply tests/test_errors.py]: The new behavior is not covered in `test_delattr` :)

File: tests/test_basic_ops_list.py [hunk 4]
```diff
@@ -738,6 +739,24 @@ def test_getattr() -> None:
     assert getattr(cfg, "2") == src[2]
 
 
+@mark.parametrize(
+    "getconfig, struct",
+    [
+        (lambda: OmegaConf.create({"name": "alice", "age": 1}), False),
+        (lambda: OmegaConf.create({"name": "alice", "age": 1}), True),
+        (lambda: OmegaConf.structured(User(name="alice", age=1)), False),
```
Comment (omry) [inline tests/test_basic_ops_list.py:747]: Why not:
```suggestion
        (OmegaConf.create({"name": "alice", "age": 1}), False),
        (OmegaConf.create({"name": "alice", "age": 1}), True),
        (OmegaConf.structured(User(name="alice", age=1)), False),
```

File: tests/test_basic_ops_list.py [hunk 5]
```diff
@@ -738,6 +739,24 @@ def test_getattr() -> None:
     assert getattr(cfg, "2") == src[2]
 
 
+@mark.parametrize(
```
Comment (omry) [inline tests/test_basic_ops_list.py:742]: `tests/test_basic_ops_list.py` seems like the wrong file for these tests.
move to test_basic_ops_dict ?

File: tests/test_basic_ops_list.py (hunk 4 above)
Comment (jieru-hu) [reply tests/test_basic_ops_list.py]: fixed!

File: tests/test_basic_ops_list.py (hunk 5 above)
Comment (jieru-hu) [reply tests/test_basic_ops_list.py]: thanks, yes, accidentally added to the wrong file :)

File: tests/test_basic_ops_dict.py [hunk 6]
```diff
@@ -75,6 +75,23 @@ def test_getattr_dict() -> None:
...
+    "cfg, struct",
+    [
+        (OmegaConf.create({"name": "alice", "age": 1}), False),
+        (OmegaConf.create({"name": "alice", "age": 1}), True),
+        (OmegaConf.structured(User(name="alice", age=1)), False),
+    ],
+)
+def test_delattr(cfg: DictConfig, struct: bool) -> None:
+    if struct:
+        OmegaConf.set_struct(cfg, True)
```
Comment (omry) [inline tests/test_basic_ops_dict.py:88]:
```suggestion
    OmegaConf.set_struct(cfg, struct)
```

File: tests/test_basic_ops_dict.py [hunk 7]
```diff
@@ -75,6 +75,23 @@ def test_getattr_dict() -> None:
...
 
 
+@mark.parametrize(
+    "cfg, struct",
+    [
+        (OmegaConf.create({"name": "alice", "age": 1}), False),
+        (OmegaConf.create({"name": "alice", "age": 1}), True),
+        (OmegaConf.structured(User(name="alice", age=1)), False),
+    ],
+)
```
Comment (omry) [inline tests/test_basic_ops_dict.py:85]: Try to parameterize like this. This will give all 4 combinations.
```python
@mark.parametrize("struct" : [True, False])
@mark.parametrize("cfg" : [
        param(OmegaConf.create({"name": "alice", "age": 1}), id="dict"),
        param(OmegaConf.structured(User(name="alice", age=1)), id="structured_config"),
])
```

Comment (jieru-hu) [reply tests/test_basic_ops_dict.py]: nice!

**Review by omry — APPROVED** [review]
shipit!
